### DISCUSSION TITLE
Failing tests

### DISCUSSION THREAD

**Discussion by emu-code** [Cyfrin/foundry-full-course-cu#2229] [General]
I'm getting this error when I try to run my test
```
Failing tests:
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: setup failed: ZeroAddress()] setUp() (gas: 0)
```

Here's my code
```
//SPDX-License-Identifier:MIT

pragma solidity 0.8.19;

import {Test} from "forge-std/Test.sol";
import {Raffle} from "src/Raffle.sol";
import {DeployRaffle} from "script/DeployRaffle.s.sol";
import {HelperConfig} from "script/HelperConfig.s.sol";

contract RaffleTest is Test {
    Raffle public raffle;
    HelperConfig public helperConfig;

    uint256 entranceFee;
    uint256 interval;
    address vrfCoordinator;
    bytes32 gasLane;
    uint256 subscriptionId;
    uint32 callBackGasLimit;

    address public PLAYER = makeAddr("player");
    uint256 public constant STARTING_PLAYER_BALANCE = 10 ether;

    function setUp() external {
        DeployRaffle deployer = new DeployRaffle();
        (raffle, helperConfig) = deployer.deployContract();

        HelperConfig.NetworkConfig memory config = helperConfig.getConfig();
        entranceFee = config.entranceFee;
        interval = config.interval;
        vrfCoordinator = config.vrfCoordinator;
        gasLane = config.gasLane;
        subscriptionId = config.subscriptionId;
        callBackGasLimit = config.callBackGasLimit;
    }

    function testRaffleInitializesInOpenState() public view {
        assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN);
    }

    // function testRaffleRevertsWithoutEnoughETH() public {
    //     // Arra
    // }
}
```

**Comment by EngrPips** [comment]
Hey @emu-code, run your code with the `-vvvv` flag and show us the test output.

**Reply by emu-code** [reply]
Here
```
Ran 1 test for test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: setup failed: ZeroAddress()] setUp() (gas: 0)
Traces:
  [11111854] RaffleTest::setUp()
    ├─ [4492114] → new DeployRaffle@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 22324 bytes of code
    ├─ [6580650] DeployRaffle::deployContract()
    │   ├─ [3467084] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   └─ ← [Return] 16530 bytes of code
    │   ├─ [3019734] HelperConfig::getConfig()
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [2863541] → new VRFCoordinatorV2_5Mock@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   ├─ emit ConfigSet()
    │   │   │   └─ ← [Return] 13951 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] NetworkConfig({ entranceFee: 0, interval: 0, vrfCoordinator: 0x0000000000000000000000000000000000000000, gasLane: 0x0000000000000000000000000000000000000000000000000000000000000000, subscriptionId: 0, callBackGasLimit: 0 })
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [22726] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   └─ ← [Revert] 4 bytes of code
    │   └─ ← [Revert] ZeroAddress()
    └─ ← [Revert] ZeroAddress()
```
Apparently my vrfCoordinator is tending to 0, but to the best of my knowledge my HelperConfig contract is as Patrick did it. Have a look
```
//SPDX-License-Identifier:MIT

pragma solidity 0.8.19;

import {Script} from "forge-std/Script.sol";
import {VRFCoordinatorV2_5Mock} from "lib/chainlink-brownie-contracts/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2_5Mock.sol";

abstract contract CodeConstants {
    /**VRF Mocks */
    uint96 public MOCK_BASE_FEE = 0.25 ether;
    uint96 public MOCK_GAS_PRICE_LINK = 1e18;
    //    LINK / ETH price
    int256 public MOCK_WEI_PER_UNIT_LINK = 4e15;

    uint256 public constant ETH_SEPOLIA_CHAIN_ID = 11155111;
    uint256 public constant LOCAL_CHAIN_ID = 31337;
}

contract HelperConfig is Script, CodeConstants {
    error NetworkConfig__InvalidChainId();
    struct NetworkConfig {
        uint256 entranceFee;
        uint256 interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint256 subscriptionId;
        uint32 callBackGasLimit;
    }

    NetworkConfig public localNetworkConfig;
    mapping(uint256 chainId => NetworkConfig) public networkConfig;

    constructor() {
        networkConfig[ETH_SEPOLIA_CHAIN_ID] = getSepoliaEthConfig();
    }

    function getConfigByChainId(
        uint256 chainId
    ) public returns (NetworkConfig memory) {
        if (networkConfig[chainId].vrfCoordinator != address(0)) {
            return networkConfig[chainId];
        } else if (chainId == LOCAL_CHAIN_ID) {
            return getOrCreateAnvilEthConfig();
        } else {
            revert NetworkConfig__InvalidChainId();
        }
    }

    function getConfig() public returns (NetworkConfig memory) {
        return getConfigByChainId(block.chainid);
    }

    function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
        return
            NetworkConfig({
                entranceFee: 0.01 ether,
                interval: 30,
                vrfCoordinator: 0x9DdfaCa8183c41ad55329BdeeD9F6A8d53168B1B,
                gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
                callBackGasLimit: 500000,
                subscriptionId: 0
            });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
        if (localNetworkConfig.vrfCoordinator != address(0)) {
            return localNetworkConfig;
        }

        //Deploy mock
        vm.startBroadcast();
        VRFCoordinatorV2_5Mock vrfCoordinatorMock = new VRFCoordinatorV2_5Mock(
            MOCK_BASE_FEE,
            MOCK_GAS_PRICE_LINK,
            MOCK_WEI_PER_UNIT_LINK
        );
        vm.stopBroadcast();

        localNetworkConfig = NetworkConfig({
            entranceFee: 0.01 ether,
            interval: 30, // 30 seconds
            vrfCoordinator: address(vrfCoordinatorMock),
            gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
            callBackGasLimit: 500000,
            subscriptionId: 0
        });
    }
}
```

**Comment by EngrPips** [comment]
Hey @emu-code, Can you check why every field in your `NetworkConfig` is `0`?

```js
├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] NetworkConfig({ entranceFee: 0, interval: 0, vrfCoordinator: 0x0000000000000000000000000000000000000000, gasLane: 0x0000000000000000000000000000000000000000000000000000000000000000, subscriptionId: 0, callBackGasLimit: 0 })
```

**Reply by emu-code** [reply]
It passes now! In the getOrCreateAnvilEthConfig() function I forgot to return localNetworkConfig()

**Reply by EngrPips** [reply]
Interesting. Keep crushing it, buddy.

**Reply by tushar1698** [reply]
I have the Same Problem Guys, even after returning the localNetworkConfig I am still getting 0 values in the network config
![image](https://github.com/user-attachments/assets/85632a34-1c12-453b-afc3-e7707ea589ab)

**Reply by EngrPips** [reply]
Run your test with the `-vvvv` flag, and then show us the test output.

**Reply by tushar1698** [reply]
![image](https://github.com/user-attachments/assets/949b4272-701b-4320-ac88-31724071e4b3)
Failing Right after the vm.stopBroadcast

**Reply by tushar1698** [reply]
@EngrPips

**Reply by EngrPips** [reply]
Henceforth, please don't share pictures; copy and paste the code here. Also, make sure to edit the code correctly. Share your `HelperConfig`

**Reply by tushar1698** [reply]
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "forge-std/Script.sol";
import "@chainlink/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2_5Mock.sol";

abstract contract CodeConstants {
    /* VRF Mock Values */
    uint96 public MOCK_BASE_FEE = 0.2 ether;
    uint96 public MOCK_GAS_PRICE = 1e9;
    // For Link/ETH price
    int256 public MOCK_PRICE_LINK = 4e15;

    uint256 public constant SEPOLIA_CHAIN_ID = 1115511;
    uint256 public constant LOCAL_CHAIN_ID = 31337;
}

contract HelperConfig is
    CodeConstants,
    Script //we need to use this as a helper so that we won't be passing constructor args at the time of deployment
{
    error HelperConfig__InvalidChainId();

    struct NetworkConfig {
        uint256 entranceFee;
        uint256 interval;
        address vrfCoordinator;
        bytes32 gasLane;
        uint256 subscriptionId;
        uint32 callbackGasLimit;
    }

    NetworkConfig public localNetworkConfig;
    mapping(uint256 chainID => NetworkConfig) public networkConfigs; // Mapping of the struct

    constructor() {
        networkConfigs[SEPOLIA_CHAIN_ID] = getSepoliaEthConfig();
    }

    function getConfigByChainId(uint256 chainID) public returns (NetworkConfig memory) {
        if (networkConfigs[chainID].vrfCoordinator != address(0)) {
            return networkConfigs[chainID];
        } else if (chainID == LOCAL_CHAIN_ID) {
            getAnvilEthConfig();
        } else {
            revert HelperConfig__InvalidChainId();
        }
    }

    function getConfig() public returns (NetworkConfig memory) {
        return getConfigByChainId(block.chainid);
    }

    function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
        return NetworkConfig({
            entranceFee: 0.01 ether, //1e16
            interval: 30, // in seconds
            vrfCoordinator: 0x9DdfaCa8183c41ad55329BdeeD9F6A8d53168B1B,
            gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
            subscriptionId: 0,
            callbackGasLimit: 500000
        });
    }

    function getAnvilEthConfig() public returns (NetworkConfig memory) {
        // check to see if we set an active network config
        if (localNetworkConfig.vrfCoordinator != address(0)) {
            return localNetworkConfig;
        }

        //Deploying Mocks
        vm.startBroadcast();
        VRFCoordinatorV2_5Mock vrfCoordinatorMock =
            new VRFCoordinatorV2_5Mock(MOCK_BASE_FEE, MOCK_GAS_PRICE, MOCK_PRICE_LINK);
            console.log(address(vrfCoordinatorMock), "<----Deployed Mock VRF Address");
        vm.stopBroadcast();

        localNetworkConfig = NetworkConfig({
            entranceFee: 0.01 ether, //1e16
            interval: 30, // in seconds
            vrfCoordinator: address(vrfCoordinatorMock),
            gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
            subscriptionId: 0,
            callbackGasLimit: 0
        });
        return localNetworkConfig;
    }
}

**Reply by EngrPips** [reply]
```js
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "forge-std/Script.sol";
import "@chainlink/contracts/src/v0.8/vrf/mocks/VRFCoordinatorV2_5Mock.sol";

abstract contract CodeConstants {
/* VRF Mock Values */
uint96 public MOCK_BASE_FEE = 0.2 ether;
uint96 public MOCK_GAS_PRICE = 1e9;
// For Link/ETH price
int256 public MOCK_PRICE_LINK = 4e15;

uint256 public constant SEPOLIA_CHAIN_ID = 1115511;
uint256 public constant LOCAL_CHAIN_ID = 31337;
}

contract HelperConfig is
CodeConstants,
Script //we need to use this as a helper so that we won't be passing constructor args at the time of deployment
{
error HelperConfig__InvalidChainId();

struct NetworkConfig {
    uint256 entranceFee;
    uint256 interval;
    address vrfCoordinator;
    bytes32 gasLane;
    uint256 subscriptionId;
    uint32 callbackGasLimit;
}

NetworkConfig public localNetworkConfig;
mapping(uint256 chainID => NetworkConfig) public networkConfigs; // Mapping of the struct

constructor() {
    networkConfigs[SEPOLIA_CHAIN_ID] = getSepoliaEthConfig();
}

function getConfigByChainId(uint256 chainID) public returns (NetworkConfig memory) {
    if (networkConfigs[chainID].vrfCoordinator != address(0)) {
        return networkConfigs[chainID];
    } else if (chainID == LOCAL_CHAIN_ID) {
        getAnvilEthConfig();
    } else {
        revert HelperConfig__InvalidChainId();
    }
}

function getConfig() public returns (NetworkConfig memory) {
    return getConfigByChainId(block.chainid);
}

function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
    return NetworkConfig({
        entranceFee: 0.01 ether, //1e16
        interval: 30, // in seconds
        vrfCoordinator: 0x9DdfaCa8183c41ad55329BdeeD9F6A8d53168B1B,
        gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
        subscriptionId: 0,
        callbackGasLimit: 500000
    });
}

function getAnvilEthConfig() public returns (NetworkConfig memory) {
    // check to see if we set an active network config
    if (localNetworkConfig.vrfCoordinator != address(0)) {
        return localNetworkConfig;
    }

    //Deploying Mocks
    vm.startBroadcast();
    VRFCoordinatorV2_5Mock vrfCoordinatorMock =
        new VRFCoordinatorV2_5Mock(MOCK_BASE_FEE, MOCK_GAS_PRICE, MOCK_PRICE_LINK);
        console.log(address(vrfCoordinatorMock), "<----Deployed Mock VRF Address");
    vm.stopBroadcast();

    localNetworkConfig = NetworkConfig({
        entranceFee: 0.01 ether, //1e16
        interval: 30, // in seconds
        vrfCoordinator: address(vrfCoordinatorMock),
        gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
        subscriptionId: 0,
        callbackGasLimit: 0
    });
    return localNetworkConfig;
}
}
```

**Reply by tushar1698** [reply]
Here is the Output @emu-code 

Traces:
  [11311808] RaffleTest::setUp()
    ├─ [4629307] → new DeployRaffle@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 23009 bytes of code
    ├─ [6644627] DeployRaffle::deployContract()
    │   ├─ [3549080] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   └─ ← [Return] 16829 bytes of code
    │   ├─ [0] console::log(HelperConfig: [0x104fBc016F4bb334D775a19E8A6510109AC63E00], "<---Deployed Helper Config by Deploy Script") [staticcall]
    │   │   └─ ← [Stop] 
    │   ├─ [2999668] HelperConfig::getConfig()
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [2863541] → new VRFCoordinatorV2_5Mock@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   ├─ emit ConfigSet()
    │   │   │   └─ ← [Return] 13951 bytes of code
    │   │   ├─ [0] console::log(VRFCoordinatorV2_5Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], "<----Deployed Mock VRF Address") [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] NetworkConfig({ entranceFee: 0, interval: 0, vrfCoordinator: 0x0000000000000000000000000000000000000000, gasLane: 0x0000000000000000000000000000000000000000000000000000000000000000, subscriptionId: 0, callbackGasLimit: 0 })
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [22726] → new <unknown>@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   └─ ← [Revert] 4 bytes of code
    │   └─ ← [Revert] ZeroAddress()
    └─ ← [Revert] ZeroAddress()

**Reply by EngrPips** [reply]
```js
Traces:
[11311808] RaffleTest::setUp()
├─ [4629307] → new DeployRaffle@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
│ └─ ← [Return] 23009 bytes of code
├─ [6644627] DeployRaffle::deployContract()
│ ├─ [3549080] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
│ │ └─ ← [Return] 16829 bytes of code
│ ├─ [0] console::log(HelperConfig: [0x104fBc016F4bb334D775a19E8A6510109AC63E00], "<---Deployed Helper Config by Deploy Script") [staticcall]
│ │ └─ ← [Stop]
│ ├─ [2999668] HelperConfig::getConfig()
│ │ ├─ [0] VM::startBroadcast()
│ │ │ └─ ← [Return]
│ │ ├─ [2863541] → new VRFCoordinatorV2_5Mock@0x90193C961A926261B756D1E5bb255e67ff9498A1
│ │ │ ├─ emit ConfigSet()
│ │ │ └─ ← [Return] 13951 bytes of code
│ │ ├─ [0] console::log(VRFCoordinatorV2_5Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], "<----Deployed Mock VRF Address") [staticcall]
│ │ │ └─ ← [Stop]
│ │ ├─ [0] VM::stopBroadcast()
│ │ │ └─ ← [Return]
│ │ └─ ← [Return] NetworkConfig({ entranceFee: 0, interval: 0, vrfCoordinator: 0x0000000000000000000000000000000000000000, gasLane: 0x0000000000000000000000000000000000000000000000000000000000000000, subscriptionId: 0, callbackGasLimit: 0 })
│ ├─ [0] VM::startBroadcast()
│ │ └─ ← [Return]
│ ├─ [22726] → new @0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
│ │ └─ ← [Revert] 4 bytes of code
│ └─ ← [Revert] ZeroAddress()
└─ ← [Revert] ZeroAddress()
```

**Reply by tushar1698** [reply]
any suggestions @emu-code

**Reply by EngrPips** [reply]
Hey @tushar1698, run `forge clean` then retry.

**Reply by tushar1698** [reply]
@emu-code It Still doesn't work, mate. Do I need to load the contract with LINK at this point of testing?

**Reply by emu-code** [reply]
In hindsight the only thing I can see is your ETH_SEPOLIA_CHAIN_ID is suppose to =  11155111;
Else copy your code on chatgpt and see whether it can spot something else out

**Reply by EngrPips** [reply]
Good catch

**Reply by EngrPips** [reply]
Show your `Deployment` script

**Reply by tushar1698** [reply]
@emu-code @EngrPips Thanks for the insight, but it has nothing to do right now as The Sepolia Chain ID is currently not in use, and that function is not being called. However, I have corrected it and cross-checked the code with the video.

**Reply by tushar1698** [reply]
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import "forge-std/Script.sol";
import "../src/Raffle.sol";
import "script/HelperConfig.s.sol";

contract DeployRaffle is Script {
    function run() public {}

    function deployContract() public returns (Raffle, HelperConfig) {
        HelperConfig helperConfig = new HelperConfig();
        console.log(address(helperConfig), "<---Deployed Helper Config by Deploy Script");

        //on Local Networks ---> Deploy Mocks --> Get Local Config
        //For Sepolia just this will just fetch the Sepolia Config
        HelperConfig.NetworkConfig memory config = helperConfig.getConfig();

        vm.startBroadcast();
        Raffle raffle = new Raffle( //The constructor params for the Raffle contract are as same as our for configs so we can just fetch configs here for deployment
            config.entranceFee,
            config.interval,
            config.vrfCoordinator,
            config.gasLane,
            config.subscriptionId,
            config.callbackGasLimit
        );
        vm.stopBroadcast();
        return (raffle, helperConfig);
        
    }
}

**Reply by emu-code** [reply]
Do tell what the problem was

**Reply by tushar1698** [reply]
@emu-code Traces:
[11311808] RaffleTest::setUp()
├─ [4629307] → new DeployRaffle@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
│ └─ ← [Return] 23009 bytes of code
├─ [6644627] DeployRaffle::deployContract()
│ ├─ [3549080] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
│ │ └─ ← [Return] 16829 bytes of code
│ ├─ [0] console::log(HelperConfig: [0x104fBc016F4bb334D775a19E8A6510109AC63E00], "<---Deployed Helper Config by Deploy Script") [staticcall]
│ │ └─ ← [Stop]
│ ├─ [2999668] HelperConfig::getConfig()
│ │ ├─ [0] VM::startBroadcast()
│ │ │ └─ ← [Return]
│ │ ├─ [2863541] → new VRFCoordinatorV2_5Mock@0x90193C961A926261B756D1E5bb255e67ff9498A1
│ │ │ ├─ emit ConfigSet()
│ │ │ └─ ← [Return] 13951 bytes of code
│ │ ├─ [0] console::log(VRFCoordinatorV2_5Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], "<----Deployed Mock VRF Address") [staticcall]
│ │ │ └─ ← [Stop]
│ │ ├─ [0] VM::stopBroadcast()
│ │ │ └─ ← [Return]
│ │ └─ ← [Return] NetworkConfig({ entranceFee: 0, interval: 0, vrfCoordinator: 0x0000000000000000000000000000000000000000, gasLane: 0x0000000000000000000000000000000000000000000000000000000000000000, subscriptionId: 0, callbackGasLimit: 0 })
│ ├─ [0] VM::startBroadcast()
│ │ └─ ← [Return]
│ ├─ [22726] → new @0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
│ │ └─ ← [Revert] 4 bytes of code
│ └─ ← [Revert] ZeroAddress()
└─ ← [Revert] ZeroAddress()

**Reply by tushar1698** [reply]
@emu-code NetworkConfig is returning zero values, Helper config contract code is posted above

**Reply by emu-code** [reply]
No, I mean since you've corrected it I wanted to know what caused it. IT's working now yeah?

**Reply by KevinFalah** [reply]
@tushar1698 I know your problem is because in your function getConfigByChainId, you not give return for getOrCreateAnvilEthConfig() when condition chainId == LOCAL_CHAIN_ID

here use this code
```
function getConfigByChainId(uint256 chainID) public returns (NetworkConfig memory) {
    if (networkConfigs[chainID].vrfCoordinator != address(0)) {
        return networkConfigs[chainID];
    } else if (chainID == LOCAL_CHAIN_ID) {
       return getAnvilEthConfig();  // <---- give return here
    } else {
        revert HelperConfig__InvalidChainId();
    }
}
```

**Reply by jaym9000** [reply]
> @emu-code NetworkConfig is returning zero values, Helper config contract code is posted above

If you're still having this issue, or anyone else, this is what fixed it for me:

Adding return localNetworkConfig in the getOrCreateAnvilEthConfig function for the HelperConfig contract


```
function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
        if (localNetworkConfig.vrfCoordinator != address(0)) {
            return localNetworkConfig;
        }

        // Deploy mocks and set up local config
        vm.startBroadcast();
        VRFCoordinatorV2_5Mock vrfCoordinatorMock = new VRFCoordinatorV2_5Mock(
            MOCK_BASE_FEE,
            MOCK_GAS_PRICE,
            MOCK_WEI_PER_UINT_LINK
        );
        vm.stopBroadcast();

        localNetworkConfig = NetworkConfig({
            entranceFee: 0.01 ether,
            interval: 30 seconds,
            vrfCoordinator: address(vrfCoordinatorMock),
            gasLane: 0x787d74caea10b2b357790d5b5247c2f63d1d91572a9846f780606e4d953677ae,
            subscriptionId: 0,
            callbackGasLimit: 500000
        });

        return localNetworkConfig;
}
```

    
And also returning raffle and helperConfig in the deployContract function of the DeployRaffle contract


```
function deployContract() public returns (Raffle, HelperConfig) {
        HelperConfig helperConfig = new HelperConfig();
        // local -> deploy mocks, get local config
        // sepolia -> get sepolia config
        HelperConfig.NetworkConfig memory config = helperConfig.getConfig();

        require(config.vrfCoordinator != address(0), "VRFCoordinator address cannot be zero");

        vm.startBroadcast();
        Raffle raffle = new Raffle(
            config.entranceFee,
            config.interval,
            config.vrfCoordinator,
            config.gasLane,
            config.subscriptionId,
            config.callbackGasLimit
        );
        vm.stopBroadcast();

        return (raffle, helperConfig);
}
```

**Reply by EngrPips** [reply]
Great heads-up here.

**Comment by EngrPips** [comment]
Interesting

**Comment by Deepizel** [comment]
for anyone coming back to this, what fixed my own issue is to just add RETURN before getOrCreateAnvilEthConfig(); when chainId == LOCAL_CHAIN_ID. just like it is below, its so easy to miss despite the fact that @jaym9000 already mentioned it 

    function getConfigByChainId(
        uint256 chainId
    ) public returns (NetworkConfig memory) {
        if (networkConfigs[chainId].vrfCoordinator != address(0)) {
            return networkConfigs[chainId];
        } else if (chainId == LOCAL_CHAIN_ID) {
           return getOrCreateAnvilEthConfig();
        } else {
            revert HelperConfig__InvalidChainId();
        }
    }

**Reply by EngrPips** [reply]
I am glad you figured it out.
